src/server/tablet_snap.rs

```diff
@@ -0,0 +1,557 @@
+    worker::Runnable,
+};
+use tokio::runtime::{Builder as RuntimeBuilder, Runtime};
+
+use super::{
+    metrics::*,
+    snap::{Task, DEFAULT_POOL_SIZE, SNAP_CHUNK_LEN},
+    Config, Error, Result,
+};
+use crate::tikv_util::{sys::thread::ThreadBuildWrapper, time::Limiter};
+
+struct RecvTabletSnapContext {
+    key: TabletSnapKey,
+    raft_msg: RaftMessage,
+    io_type: IoType,
+    start: Instant,
+    chunk_size: usize,
+}
+
+impl RecvTabletSnapContext {
+    fn new(mut head: SnapshotChunk) -> Result<Self> {
+        if !head.has_message() {
+            return Err(box_err!("no raft message in the first chunk"));
+        }
+
+        let chunk_size = match head.take_data().try_into() {
+            Ok(buff) => usize::from_ne_bytes(buff),
+            Err(_) => return Err(box_err!("failed to get chunk size")),
+        };
+        let meta = head.take_message();
+        let key = TabletSnapKey::from_region_snap(
+            meta.get_region_id(),
+            meta.get_to_peer().get_id(),
+            meta.get_message().get_snapshot(),
+        );
+        let io_type = io_type_from_raft_message(&meta)?;
+
+        Ok(RecvTabletSnapContext {
+            key,
+            raft_msg: meta,
+            io_type,
+            start: Instant::now(),
+            chunk_size,
+        })
+    }
+
+    fn finish<R: RaftStoreRouter<impl KvEngine>>(self, raft_router: R) -> Result<()> {
+        let key = self.key;
+        if let Err(e) = raft_router.send_raft_msg(self.raft_msg) {
+            return Err(box_err!("{} failed to send snapshot to raft: {}", key, e));
+        }
+        info!("saving all snapshot files"; "snap_key" => %key, "takes" => ?self.start.saturating_elapsed());
+        Ok(())
+    }
+}
+
+fn io_type_from_raft_message(msg: &RaftMessage) -> Result<IoType> {
+    let snapshot = msg.get_message().get_snapshot();
+    let data = snapshot.get_data();
+    let mut snapshot_data = RaftSnapshotData::default();
+    snapshot_data.merge_from_bytes(data)?;
+    let snapshot_meta = snapshot_data.get_meta();
+    if snapshot_meta.get_for_balance() {
+        Ok(IoType::LoadBalance)
+    } else {
+        Ok(IoType::Replication)
+    }
+}
+
+async fn send_snap_files(
+    mgr: &TabletSnapManager,
+    mut sender: impl Sink<(SnapshotChunk, WriteFlags), Error = Error> + Unpin,
+    msg: RaftMessage,
+    key: TabletSnapKey,
+    limiter: Limiter,
+) -> Result<u64> {
+    let path = mgr.get_final_path_for_gen(&key);
+    info!("begin to send snapshot file";"snap_key" => %key);
+    let files = fs::read_dir(&path)?
+        .map(|f| Ok(f?.path()))
+        .filter(|f| f.is_ok() && f.as_ref().unwrap().is_file())
+        .collect::<Result<Vec<_>>>()?;
+    let io_type = io_type_from_raft_message(&msg)?;
+    let _with_io_type = WithIoType::new(io_type);
+    let mut total_sent = msg.compute_size() as u64;
+    let mut chunk = SnapshotChunk::default();
+    chunk.set_message(msg);
+    chunk.set_data(usize::to_ne_bytes(SNAP_CHUNK_LEN).to_vec());
```

Comment:
It's confusing to set both fields at the same time.

Reply:
The snapchunk size can be changed in later version. It can keep compatibility.